In [6]:
import requests
url = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text

In [7]:
!pip install BeautifulSoup4
from bs4 import BeautifulSoup
soup = BeautifulSoup(url,'html.parser')
print(soup.prettify())

<!DOCTYPE html>
<html class="client-nojs" dir="ltr" lang="en">
 <head>
  <meta charset="utf-8"/>
  <title>
   List of postal codes of Canada: M - Wikipedia
  </title>
  <script>
   document.documentElement.className="client-js";RLCONF={"wgBreakFrames":!1,"wgSeparatorTransformTable":["",""],"wgDigitTransformTable":["",""],"wgDefaultDateFormat":"dmy","wgMonthNames":["","January","February","March","April","May","June","July","August","September","October","November","December"],"wgRequestId":"619aa236-74a4-4d93-b682-88d9dbe449b8","wgCSPNonce":!1,"wgCanonicalNamespace":"","wgCanonicalSpecialPageName":!1,"wgNamespaceNumber":0,"wgPageName":"List_of_postal_codes_of_Canada:_M","wgTitle":"List of postal codes of Canada: M","wgCurRevisionId":960187814,"wgRevisionId":960187814,"wgArticleId":539066,"wgIsArticle":!0,"wgIsRedirect":!1,"wgAction":"view","wgUserName":null,"wgUserGroups":["*"],"wgCategories":["Articles with short description","Communications in Ontario","Postal codes in Canada","Toron

In [8]:
tb = soup.find('table')
fields = tb.find_all('td')

postcode = []
borough = []
neighbourhood = []

for i in range(0, len(fields), 3):
    postcode.append(fields[i].text.strip())
    borough.append(fields[i+1].text.strip())
    neighbourhood.append(fields[i+2].text.strip())
        
DFP = pd.DataFrame(data=[postcode, borough, neighbourhood]).transpose()
DFP.columns = ['Postcode', 'Borough', 'Neighbourhood']
DFP.head()

Postcode           Borough              Neighbourhood
0      M1A      Not assigned               Not assigned
1      M2A      Not assigned               Not assigned
2      M3A        North York                  Parkwoods
3      M4A        North York           Victoria Village
4      M5A  Downtown Toronto  Regent Park, Harbourfront

In [9]:
DFP['Borough'].replace('Not assigned', np.nan, inplace=True)
DFP.dropna(subset=['Borough'], inplace=True)

DFP.head()


Postcode           Borough                                Neighbourhood
2      M3A        North York                                    Parkwoods
3      M4A        North York                             Victoria Village
4      M5A  Downtown Toronto                    Regent Park, Harbourfront
5      M6A        North York             Lawrence Manor, Lawrence Heights
6      M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [10]:
DFP1 = DFP.groupby(['Postcode', 'Borough'])['Neighbourhood'].apply(', '.join).reset_index()
DFP1.columns = ['Postcode', 'Borough', 'Neighbourhood']
DFP1

Postcode           Borough  \
0        M1B       Scarborough   
1        M1C       Scarborough   
2        M1E       Scarborough   
3        M1G       Scarborough   
4        M1H       Scarborough   
5        M1J       Scarborough   
6        M1K       Scarborough   
7        M1L       Scarborough   
8        M1M       Scarborough   
9        M1N       Scarborough   
10       M1P       Scarborough   
11       M1R       Scarborough   
12       M1S       Scarborough   
13       M1T       Scarborough   
14       M1V       Scarborough   
15       M1W       Scarborough   
16       M1X       Scarborough   
17       M2H        North York   
18       M2J        North York   
19       M2K        North York   
20       M2L        North York   
21       M2M        North York   
22       M2N        North York   
23       M2P        North York   
24       M2R        North York   
25       M3A        North York   
26       M3B        North York   
27       M3C        North York   
28       M3H        North York   
29       M3J        North York   
30       M3K        North York   
31       M3L        North York   
32       M3M        North York   
33       M3N        North York   
34       M4A        North York   
35       M4B         East York   
36       M4C         East York   
37       M4E      East Toronto   
38       M4G         East York   
39       M4H         East York   
40       M4J         East York   
41       M4K      East Toronto   
42       M4L      East Toronto   
43       M4M      East Toronto   
44       M4N   Central Toronto   
45       M4P   Central Toronto   
46       M4R   Central Toronto   
47       M4S   Central Toronto   
48       M4T   Central Toronto   
49       M4V   Central Toronto   
50       M4W  Downtown Toronto   
51       M4X  Downtown Toronto   
52       M4Y  Downtown Toronto   
53       M5A  Downtown Toronto   
54       M5B  Downtown Toronto   
55       M5C  Downtown Toronto   
56       M5E  Downtown Toronto   
57       M5G  Downtown Toronto   
58       M5H  Downtown Toronto   
59       M5J  Downtown Toronto   
60       M5K  Downtown Toronto   
61       M5L  Downtown Toronto   
62       M5M        North York   
63       M5N   Central Toronto   
64       M5P   Central Toronto   
65       M5R   Central Toronto   
66       M5S  Downtown Toronto   
67       M5T  Downtown Toronto   
68       M5V  Downtown Toronto   
69       M5W  Downtown Toronto   
70       M5X  Downtown Toronto   
71       M6A        North York   
72       M6B        North York   
73       M6C              York   
74       M6E              York   
75       M6G  Downtown Toronto   
76       M6H      West Toronto   
77       M6J      West Toronto   
78       M6K      West Toronto   
79       M6L        North York   
80       M6M              York   
81       M6N              York   
82       M6P      West Toronto   
83       M6R      West Toronto   
84       M6S      West Toronto   
85       M7A  Downtown Toronto   
86       M7R       Mississauga   
87       M7Y      East Toronto   
88       M8V         Etobicoke   
89       M8W         Etobicoke   
90       M8X         Etobicoke   
91       M8Y         Etobicoke   
92       M8Z         Etobicoke   
93       M9A         Etobicoke   
94       M9B         Etobicoke   
95       M9C         Etobicoke   
96       M9L        North York   
97       M9M        North York   
98       M9N              York   
99       M9P         Etobicoke   
100      M9R         Etobicoke   
101      M9V         Etobicoke   
102      M9W         Etobicoke   

                                         Neighbourhood  
0                                       Malvern, Rouge  
1               Rouge Hill, Port Union, Highland Creek  
2                    Guildwood, Morningside, West Hill  
3                                               Woburn  
4                                            Cedarbrae  
5                                  Scarborough Village  
6          Kennedy Park, Ionview, East Birchmount Park  
7          

I am going to assign the park: Yankee, to the not-assigned neighbours

In [11]:
DFP1['Neighbourhood'].replace('Not assigned', "Yankee", inplace=True)

DFP1

Postcode           Borough  \
0        M1B       Scarborough   
1        M1C       Scarborough   
2        M1E       Scarborough   
3        M1G       Scarborough   
4        M1H       Scarborough   
5        M1J       Scarborough   
6        M1K       Scarborough   
7        M1L       Scarborough   
8        M1M       Scarborough   
9        M1N       Scarborough   
10       M1P       Scarborough   
11       M1R       Scarborough   
12       M1S       Scarborough   
13       M1T       Scarborough   
14       M1V       Scarborough   
15       M1W       Scarborough   
16       M1X       Scarborough   
17       M2H        North York   
18       M2J        North York   
19       M2K        North York   
20       M2L        North York   
21       M2M        North York   
22       M2N        North York   
23       M2P        North York   
24       M2R        North York   
25       M3A        North York   
26       M3B        North York   
27       M3C        North York   
28       M3H        North York   
29       M3J        North York   
30       M3K        North York   
31       M3L        North York   
32       M3M        North York   
33       M3N        North York   
34       M4A        North York   
35       M4B         East York   
36       M4C         East York   
37       M4E      East Toronto   
38       M4G         East York   
39       M4H         East York   
40       M4J         East York   
41       M4K      East Toronto   
42       M4L      East Toronto   
43       M4M      East Toronto   
44       M4N   Central Toronto   
45       M4P   Central Toronto   
46       M4R   Central Toronto   
47       M4S   Central Toronto   
48       M4T   Central Toronto   
49       M4V   Central Toronto   
50       M4W  Downtown Toronto   
51       M4X  Downtown Toronto   
52       M4Y  Downtown Toronto   
53       M5A  Downtown Toronto   
54       M5B  Downtown Toronto   
55       M5C  Downtown Toronto   
56       M5E  Downtown Toronto   
57       M5G  Downtown Toronto   
58       M5H  Downtown Toronto   
59       M5J  Downtown Toronto   
60       M5K  Downtown Toronto   
61       M5L  Downtown Toronto   
62       M5M        North York   
63       M5N   Central Toronto   
64       M5P   Central Toronto   
65       M5R   Central Toronto   
66       M5S  Downtown Toronto   
67       M5T  Downtown Toronto   
68       M5V  Downtown Toronto   
69       M5W  Downtown Toronto   
70       M5X  Downtown Toronto   
71       M6A        North York   
72       M6B        North York   
73       M6C              York   
74       M6E              York   
75       M6G  Downtown Toronto   
76       M6H      West Toronto   
77       M6J      West Toronto   
78       M6K      West Toronto   
79       M6L        North York   
80       M6M              York   
81       M6N              York   
82       M6P      West Toronto   
83       M6R      West Toronto   
84       M6S      West Toronto   
85       M7A  Downtown Toronto   
86       M7R       Mississauga   
87       M7Y      East Toronto   
88       M8V         Etobicoke   
89       M8W         Etobicoke   
90       M8X         Etobicoke   
91       M8Y         Etobicoke   
92       M8Z         Etobicoke   
93       M9A         Etobicoke   
94       M9B         Etobicoke   
95       M9C         Etobicoke   
96       M9L        North York   
97       M9M        North York   
98       M9N              York   
99       M9P         Etobicoke   
100      M9R         Etobicoke   
101      M9V         Etobicoke   
102      M9W         Etobicoke   

                                         Neighbourhood  
0                                       Malvern, Rouge  
1               Rouge Hill, Port Union, Highland Creek  
2                    Guildwood, Morningside, West Hill  
3                                               Woburn  
4                                            Cedarbrae  
5                                  Scarborough Village  
6          Kennedy Park, Ionview, East Birchmount Park  
7          

In [12]:
DFP1.shape

(103, 3)

In [13]:
import pandas as pd


In [14]:
DFG = pd.read_csv('http://cocl.us/Geospatial_data')
DFG.columns = ['Postcode', 'Latitude', 'Longitude']

In [15]:
DFPS = pd.merge(DFP1, DFG, on=['Postcode'], how='inner')

DFT = DFPS[['Borough', 'Neighbourhood', 'Postcode', 'Latitude', 'Longitude']].copy()

DFT.head(12)

Borough                                      Neighbourhood Postcode  \
0   Scarborough                                     Malvern, Rouge      M1B   
1   Scarborough             Rouge Hill, Port Union, Highland Creek      M1C   
2   Scarborough                  Guildwood, Morningside, West Hill      M1E   
3   Scarborough                                             Woburn      M1G   
4   Scarborough                                          Cedarbrae      M1H   
5   Scarborough                                Scarborough Village      M1J   
6   Scarborough        Kennedy Park, Ionview, East Birchmount Park      M1K   
7   Scarborough                    Golden Mile, Clairlea, Oakridge      M1L   
8   Scarborough    Cliffside, Cliffcrest, Scarborough Village West      M1M   
9   Scarborough                        Birch Cliff, Cliffside West      M1N   
10  Scarborough  Dorset Park, Wexford Heights, Scarborough Town...      M1P   
11  Scarborough                                  Wexford, Maryvale      M1R   

     Latitude  Longitude  
0   43.806686 -79.194353  
1   43.784535 -79.160497  
2   43.763573 -79.188711  
3   43.770992 -79.216917  
4   43.773136 -79.239476  
5   43.744734 -79.239476  
6   43.727929 -79.262029  
7   43.711112 -79.284577  
8   43.716316 -79.239476  
9   43.692657 -79.264848  
10  43.757410 -79.273304  
11  43.750072 -79.295849

In [16]:
address = 'Toronto, Canada'

geolocator = Nominatim()
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of the City of Toronto are {}, {}.'.format(latitude, longitude))


/opt/conda/envs/Python36/lib/python3.6/site-packages/ipykernel/__main__.py:3: DeprecationWarning: Using Nominatim with the default "geopy/1.18.1" `user_agent` is strongly discouraged, as it violates Nominatim's ToS https://operations.osmfoundation.org/policies/nominatim/ and may possibly cause 403 and 429 HTTP errors. Please specify a custom `user_agent` with `Nominatim(user_agent="my-application")` or by overriding the default `user_agent`: `geopy.geocoders.options.default_user_agent = "my-application"`. In geopy 2.0 this will become an exception.
  app.launch_new_instance()


The geograpical coordinate of the City of Toronto are 43.6534817, -79.3839347.


In [17]:
# create map of Toronto using latitude and longitude values
MT = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(DFT['Latitude'], DFT['Longitude'], DFT['Borough'], DFT['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=3,
        popup=label,
        color='green',
        fill=True,
        fill_color='#3199cc',
        fill_opacity=0.3,
        parse_html=False).add_to(MT)  
    
MT

In [18]:
CLIENT_ID = 'SRRESTAREHIDDEN5O' # your Foursquare ID
CLIENT_SECRET = 'L3RESTAREHIDDENDC' # your Foursquare Secret
VERSION = '20200626' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: SRRESTAREHIDDEN5O
CLIENT_SECRET:L3RESTAREHIDDENDC


In [19]:
DFT1 = DFT[DFT['Borough'].str.contains('Toronto')]

TD = DFT1.reset_index(drop=True)
TD

Borough                                      Neighbourhood  \
0       East Toronto                                        The Beaches   
1       East Toronto                       The Danforth West, Riverdale   
2       East Toronto                     India Bazaar, The Beaches West   
3       East Toronto                                    Studio District   
4    Central Toronto                                      Lawrence Park   
5    Central Toronto                                   Davisville North   
6    Central Toronto                 North Toronto West,  Lawrence Park   
7    Central Toronto                                         Davisville   
8    Central Toronto                        Moore Park, Summerhill East   
9    Central Toronto  Summerhill West, Rathnelly, South Hill, Forest...   
10  Downtown Toronto                                           Rosedale   
11  Downtown Toronto                        St. James Town, Cabbagetown   
12  Downtown Toronto                               Church and Wellesley   
13  Downtown Toronto                          Regent Park, Harbourfront   
14  Downtown Toronto                           Garden District, Ryerson   
15  Downtown Toronto                                     St. James Town   
16  Downtown Toronto                                        Berczy Park   
17  Downtown Toronto                                 Central Bay Street   
18  Downtown Toronto                           Richmond, Adelaide, King   
19  Downtown Toronto  Harbourfront East, Union Station, Toronto Islands   
20  Downtown Toronto           Toronto Dominion Centre, Design Exchange   
21  Downtown Toronto                     Commerce Court, Victoria Hotel   
22   Central Toronto                                           Roselawn   
23   Central Toronto    Forest Hill North & West, Forest Hill Road Park   
24   Central Toronto                The Annex, North Midtown, Yorkville   
25  Downtown Toronto                     University of Toronto, Harbord   
26  Downtown Toronto          Kensington Market, Chinatown, Grange Park   
27  Downtown Toronto  CN Tower, King and Spadina, Railway Lands, Har...   
28  Downtown Toronto                                     Stn A PO Boxes   
29  Downtown Toronto             First Canadian Place, Underground city   
30  Downtown Toronto                                           Christie   
31      West Toronto                       Dufferin, Dovercourt Village   
32      West Toronto                           Little Portugal, Trinity   
33      West Toronto       Brockton, Parkdale Village, Exhibition Place   
34      West Toronto                      High Park, The Junction South   
35      West Toronto                             Parkdale, Roncesvalles   
36      West Toronto                                 Runnymede, Swansea   
37  Downtown Toronto        Queen's Park, Ontario Provincial Government   
38      East Toronto  Business reply mail Processing Centre, South C...   

   Postcode   Latitude  Longitude  
0       M4E  43.676357 -79.293031  
1       M4K  43.679557 -79.352188  
2       M4L  43.668999 -79.315572  
3       M4M  43.659526 -79.340923  
4       M4N  43.728020 -79.388790  
5       M4P  43.712751 -79.390197  
6       M4R  43.715383 -79.405678  
7       M4S  43.704324 -79.388790  
8       M4T  43.689574 -79.383160  
9       M4V  43.686412 -79.400049  
10      M4W  43.679563 -79.377529  
11      M4X  43.667967 -79.367675  
12      M4Y  43.665860 -79.383160  
13      M5A  43.654260 -79.360636  
14      M5B  43.657162 -79.378937  
15      M5C  43.651494 -79.375418  
16      M5E  43.644771 -79.373306  
17      M5G  43.657952 -79.387383  
18      M5H  43.650571 -79.384568  
19      M5J  43.640816 -79.381752  
20      M5K  43.647177 -79.381576  
21      M5L  43.648198 -79.379817  
22      M5N  43.711695 -79.416936  
23      M5P  43.696948 -79.411307  
24      M5R  43.672710 -79.405678  
25      M5S  43.662696 -79.400049  
26      M5T  43.653206 -79.400049  
27  

<h2>Toronto neighbourhoods</h2>

In [20]:
TM = folium.Map(location=[latitude, longitude], zoom_start=10)


for lat, lng, borough, neighborhood in zip(TD['Latitude'], TD['Longitude'], TD['Borough'], TD['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=3,
        popup=label,
        color='green',
        fill=True,
        fill_color='#3199cc',
        fill_opacity=0.3,
        parse_html=False).add_to(TM)  
    
TM

<h2>Trying to explore the very First neighborhood</h2>

In [21]:
TD.loc[0, 'Neighbourhood']

'The Beaches'

In [22]:
neighbourhood_latitude = TD.loc[0, 'Latitude'] 
neighbourhood_longitude = TD.loc[0, 'Longitude'] # neighbourhood longitude value

neighbourhood_name = TD.loc[0, 'Neighbourhood'] # neighbourhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighbourhood_name, 
                                                               neighbourhood_latitude, 
                                                               neighbourhood_longitude))


Latitude and longitude values of The Beaches are 43.67635739999999, -79.2930312.


In [23]:
LIMIT  = 80
radius = 450

url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighbourhood_latitude, 
    neighbourhood_longitude, 
    radius, 
    LIMIT)
url

'https://api.foursquare.com/v2/venues/explore?&client_id=SRRESTAREHIDDEN5O&client_secret=L3RESTAREHIDDENDC&v=20200626&ll=43.67635739999999,-79.2930312&radius=450&limit=80'

In [24]:
RS = requests.get(url).json()
RS

{'meta': {'code': 400,
  'errorType': 'invalid_auth',
  'errorDetail': 'Missing access credentials. See https://developer.foursquare.com/docs/api/configuration/authentication for details.',
  'requestId': '5efa73e828da6479402afd28'},
 'response': {}}

<h2>function that extracts the category of the venue</h2>


In [25]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

<h2>Cleaning json and structuring into Pandas Data Frame</h2>

<h1> Now I go to the task for week 5 work</h1>

<h1> Segmenting and Clustering Neighborhoods. I will explore the town; Fredericton</h1>

Fredericton is the capital city of the Canadian Province of New Brunswick. The city is situated in the west-central portion of the province along the Saint John River, 
which flows west to east as it bisects the city. By the 2016 Canadian Census, the city had a population of 58,220. It is the third-largest city in the province after Moncton 
and Saint John. There are a whole lot of venues and other attractive places in the town including government agencies,
two universities. With the number of attractive places and government agencies, there is the need to do some quantitative explorations about some venues within the city.

<h1> Data to use</h1>
To be able to explore some venues in the city, we will have to get some data about the city.

Open Data Site: http://data-fredericton.opendata.arcgis.com/

Fredericton Neighbourhoods: http://data-fredericton.opendata.arcgis.com/datasets/neighbourhoods--quartiers

Fredericton Crime by Neighbourhood: http://data-fredericton.opendata.arcgis.com/datasets/crime-by-neighbourhood-2017--crime-par-quartier-2017

Fredericton Census Tract Demographics: http://data-fredericton.opendata.arcgis.com/datasets/census-tract-demographics--donn%C3%A9es-d%C3%A9mographiques-du-secteur-de-recensement

Fredericton locations of interest: https://github.com/JasonLUrquhart/Applied-Data-Science-Capstone/blob/master/Fredericton%20Locations.xlsx

Foursquare Developers Access to venue data: https://foursquare.com/

<h1> Data Analysis</h1>

I am loading libraries that will be needed in the work

In [28]:
import numpy as np

import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json

!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim 

import requests 
from pandas.io.json import json_normalize 


import matplotlib.cm as cm
import matplotlib.colors as colors

#I am going to use k-means 
from sklearn.cluster import KMeans

 
from bs4 import BeautifulSoup

import xml

!conda install -c conda-forge folium=0.5.0 --yes 
import folium 

print('Libraries imported.')

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    geopy-2.0.0                |     pyh9f0ad1d_0          63 KB  conda-forge
    geographiclib-1.50         |             py_0          34 KB  conda-forge
    ------------------------------------------------------------
                                           Total:          97 KB

The following NEW packages will be INSTALLED:

    geographiclib: 1.50-py_0          conda-forge
    geopy:         2.0.0-pyh9f0ad1d_0 conda-forge


geopy-2.0.0          | 63 KB     | ##################################### | 100% 
geographiclib-1.50   | 34 KB     | ##################################### | 100% 
Preparing transaction: done
Verifying transaction: done
Executing transaction: done
Solving environment: done

# All requ

In [29]:
D = requests.get('https://opendata.arcgis.com/datasets/3b3c641af58e4acfbee5603a34297ccf_0.geojson')
FG = D.json()

In [30]:
FG

{'type': 'FeatureCollection',
 'name': 'Neighbourhoods___Quartiers',
 'crs': {'type': 'name',
  'properties': {'name': 'urn:ogc:def:crs:OGC:1.3:CRS84'}},
 'features': [{'type': 'Feature',
   'properties': {'OBJECTID': 1,
    'Neighbourhood': 'Fredericton South',
    'Shape__Area': 32432807.6835938,
    'Shape__Length': 40411.5512530081},
   'geometry': {'type': 'Polygon',
    'coordinates': [[[-66.6805565525253, 45.945658531517],
      [-66.6790403015066, 45.9450115443884],
      [-66.678257639474, 45.9446767197465],
      [-66.677305155268, 45.9442692754024],
      [-66.6758761306642, 45.9436574195795],
      [-66.6745833744504, 45.9431006709014],
      [-66.6734088048215, 45.9425932132558],
      [-66.6721314811295, 45.942047527692],
      [-66.6717018136486, 45.94186249269],
      [-66.6715811138511, 45.9418155705943],
      [-66.6714574138913, 45.9417726180596],
      [-66.6713309835781, 45.941733728243],
      [-66.6712020952717, 45.9416989862012],
      [-66.671168138626, 45.9416

In [31]:

G = requests.get('https://opendata.arcgis.com/datasets/6179d35eacb144a5b5fdcc869f86dfb5_0.geojson')
DG = G.json()

In [32]:
DG

{'type': 'FeatureCollection',
 'name': '9506612b-1637-4a61-a280-dbb157391d632020330-1-1pnlf56.jvrdg',
 'crs': {'type': 'name',
  'properties': {'name': 'urn:ogc:def:crs:OGC:1.3:CRS84'}},
 'features': [{'type': 'Feature',
   'properties': {'FID': 1,
    'OBJECTID': 501,
    'DBUID': '1310024304',
    'DAUID': '13100243',
    'CDUID': '1310',
    'CTUID': '3200002.00',
    'CTNAME': '0002.00',
    'DBuid_1': '1310024304',
    'DBpop2011': 60,
    'DBtdwell20': 25,
    'DBurdwell2': 22,
    'Shape_Leng': 0.00746165241824,
    'Shape_Area': 2.81310751889e-06,
    'CTIDLINK': 3200002,
    'Shape__Area': 50139.15234375,
    'Shape__Length': 929.421025287179},
   'geometry': {'type': 'Polygon',
    'coordinates': [[[-66.634784212921, 45.9519239912381],
      [-66.6351046935752, 45.9507605156138],
      [-66.6378263667982, 45.9510868696778],
      [-66.636944377136, 45.9521037018384],
      [-66.634784212921, 45.9519239912381]]]}},
  {'type': 'Feature',
   'properties': {'FID': 2,
    'OBJECTI

In [33]:
import os

In [34]:
os.listdir('.')

[]

In [35]:
OC = requests.get('https://opendata.arcgis.com/datasets/0ff4acd0a2a14096984f85c06fe4e38e_0.geojson')

In [36]:
OPD = OC.json()

In [37]:
OPD

{'type': 'FeatureCollection',
 'name': 'Crime_by_neighbourhood_2017___Crime_par_quartier_2017',
 'crs': {'type': 'name',
  'properties': {'name': 'urn:ogc:def:crs:OGC:1.3:CRS84'}},
 'features': [{'type': 'Feature',
   'properties': {'Neighbourhood': 'Fredericton South',
    'From_Date': '2017/01/05 00:00:00+00',
    'To_Date': '2017/01/26 00:00:00+00',
    'Crime_Code': 2120,
    'Crime_Type': 'B&E NON-RESIDNCE',
    'Ward': 7,
    'City': 'Fredericton',
    'FID': 1},
   'geometry': None},
  {'type': 'Feature',
   'properties': {'Neighbourhood': 'Fredericton South',
    'From_Date': '2017/03/04 00:00:00+00',
    'To_Date': '2017/03/06 00:00:00+00',
    'Crime_Code': 2120,
    'Crime_Type': 'B&E NON-RESIDNCE',
    'Ward': 7,
    'City': 'Fredericton',
    'FID': 2},
   'geometry': None},
  {'type': 'Feature',
   'properties': {'Neighbourhood': 'Fredericton South',
    'From_Date': '2017/05/07 00:00:00+00',
    'To_Date': '1970/01/01 00:00:00+00',
    'Crime_Code': 2120,
    'Crime_Type

In [39]:
# The code was removed by Watson Studio for sharing.

Neighbourhood               From_Date                 To_Date  \
0  Fredericton South  2017/01/05 00:00:00+00  2017/01/26 00:00:00+00   
1  Fredericton South  2017/03/04 00:00:00+00  2017/03/06 00:00:00+00   
2  Fredericton South  2017/05/07 00:00:00+00  1970/01/01 00:00:00+00   
3  Fredericton South  2017/06/20 00:00:00+00  2017/06/21 00:00:00+00   
4  Fredericton South  2017/07/09 00:00:00+00  2017/07/10 00:00:00+00   

   Crime_Code        Crime_Type  Ward         City  FID  
0        2120  B&E NON-RESIDNCE     7  Fredericton    1  
1        2120  B&E NON-RESIDNCE     7  Fredericton    2  
2        2120  B&E NON-RESIDNCE    12  Fredericton    3  
3        2120  B&E NON-RESIDNCE    12  Fredericton    4  
4        2120  B&E NON-RESIDNCE     7  Fredericton    5

<h3> I am going to drop the dates because they are not going to be used in this work</h3>

In [40]:
df_data_1.drop(['From_Date','To_Date'], axis = 1, inplace = True)

<h3> I want to check crime count by neighborhood and name the number as Count</h3>

In [41]:
CD = df_data_1.groupby(['Neighbourhood']).size().to_frame(name = 'Count').reset_index()

In [42]:
CD

Neighbourhood  Count
0                            Barkers Point     47
1                                Brookside     54
2                        Brookside Estates      9
3                 Brookside Mini Home Park      5
4                             College Hill     41
5                         Colonial heights      9
6                        Cotton Mill Creek      4
7                           Diamond Street      1
8                                Doak Road      1
9                                  Douglas      3
10                                Downtown    127
11                          Dun's Crossing     18
12                             Forest Hill     12
13                       Fredericton South     85
14                          Fulton Heights     36
15                            Garden Creek     13
16                            Garden Place      4
17                        Gilridge Estates      3
18                               Golf Club      7
19                           Grasse Circle      1
20                 Greenwood Minihome Park      2
21                           Hanwell North      8
22                           Heron Springs      3
23                         Highpoint Ridge      5
24             Kelly's Court Minihome Park      1
25                               Knob Hill      4
26                          Knowledge Park      1
27                          Lian / Valcore      7
28                                 Lincoln     13
29                         Lincoln Heights     14
30                             Main Street     78
31                              Marysville     39
32                                McKnight      4
33                             McLeod Hill      3
34                     Monteith / Talisman     12
35             Montogomery / Prospect East     16
36                             Nashwaaksis     25
37                 Nethervue Minihome Park      1
38                             North Devon    113
39                      Northbrook Heights     10
40                                   Platt    198
41                             Poet's Hill      4
42                                Prospect     81
43                               Rail Side      3
44                          Regiment Creek      1
45                              Royal Road      7
46               Saint Mary's First Nation     25
47                 Saint Thomas University      1
48                              Sandyville      9
49                           Serenity Lane      2
50                       Shadowood Estates      5
51                              Silverwood     12
52                           Skyline Acrea     27
53                             South Devon     68
54                          Southwood Park     16
55                              Springhill      1
56                        Sunshine Gardens     10
57                                The Hill     44
58  The Hugh John Flemming Forestry Center      3
59             University Of New Brunswick     15
60                            Waterloo Row      9
61                          Wesbett / Case      1
62                              West Hills      5
63                 Williams / Hawkins Area     17
64                          Woodstock Road     41
65                         Youngs Crossing     16

In [43]:
CD.describe()

Count
count   66.000000
mean    22.121212
std     34.879359
min      1.000000
25%      3.000000
50%      9.000000
75%     23.250000
max    198.000000

In [44]:
CD.rename(index = str, columns = {'Neighbourhood': 'Neighb', 'Count': 'C_Count'}, inplace = True)

In [45]:
CD

Neighb  C_Count
0                            Barkers Point       47
1                                Brookside       54
2                        Brookside Estates        9
3                 Brookside Mini Home Park        5
4                             College Hill       41
5                         Colonial heights        9
6                        Cotton Mill Creek        4
7                           Diamond Street        1
8                                Doak Road        1
9                                  Douglas        3
10                                Downtown      127
11                          Dun's Crossing       18
12                             Forest Hill       12
13                       Fredericton South       85
14                          Fulton Heights       36
15                            Garden Creek       13
16                            Garden Place        4
17                        Gilridge Estates        3
18                               Golf Club        7
19                           Grasse Circle        1
20                 Greenwood Minihome Park        2
21                           Hanwell North        8
22                           Heron Springs        3
23                         Highpoint Ridge        5
24             Kelly's Court Minihome Park        1
25                               Knob Hill        4
26                          Knowledge Park        1
27                          Lian / Valcore        7
28                                 Lincoln       13
29                         Lincoln Heights       14
30                             Main Street       78
31                              Marysville       39
32                                McKnight        4
33                             McLeod Hill        3
34                     Monteith / Talisman       12
35             Montogomery / Prospect East       16
36                             Nashwaaksis       25
37                 Nethervue Minihome Park        1
38                             North Devon      113
39                      Northbrook Heights       10
40                                   Platt      198
41                             Poet's Hill        4
42                                Prospect       81
43                               Rail Side        3
44                          Regiment Creek        1
45                              Royal Road        7
46               Saint Mary's First Nation       25
47                 Saint Thomas University        1
48                              Sandyville        9
49                           Serenity Lane        2
50                       Shadowood Estates        5
51                              Silverwood       12
52                           Skyline Acrea       27
53                             South Devon       68
54                          Southwood Park       16
55                              Springhill        1
56                        Sunshine Gardens       10
57                                The Hill       44
58  The Hugh John Flemming Forestry Center        3
59             University Of New Brunswick       15
60                            Waterloo Row        9
61                          Wesbett / Case        1
62                              West Hills        5
63                 Williams / Hawkins Area       17
64                          Woodstock Road       41
65                         Youngs Crossing       16

In [54]:
CD.rename({'Platt': 'PLT'}, inplace = True)
CD.rename(index=str, columns={'Neighbourhood':'Neighb','Count':'C_Count'}, inplace=True)
CD

Neighb  C_Count
0                            Barkers Point       47
1                                Brookside       54
2                        Brookside Estates        9
3                 Brookside Mini Home Park        5
4                             College Hill       41
5                         Colonial heights        9
6                        Cotton Mill Creek        4
7                           Diamond Street        1
8                                Doak Road        1
9                                  Douglas        3
10                                Downtown      127
11                          Dun's Crossing       18
12                             Forest Hill       12
13                       Fredericton South       85
14                          Fulton Heights       36
15                            Garden Creek       13
16                            Garden Place        4
17                        Gilridge Estates        3
18                               Golf Club        7
19                           Grasse Circle        1
20                 Greenwood Minihome Park        2
21                           Hanwell North        8
22                           Heron Springs        3
23                         Highpoint Ridge        5
24             Kelly's Court Minihome Park        1
25                               Knob Hill        4
26                          Knowledge Park        1
27                          Lian / Valcore        7
28                                 Lincoln       13
29                         Lincoln Heights       14
30                             Main Street       78
31                              Marysville       39
32                                McKnight        4
33                             McLeod Hill        3
34                     Monteith / Talisman       12
35             Montogomery / Prospect East       16
36                             Nashwaaksis       25
37                 Nethervue Minihome Park        1
38                             North Devon      113
39                      Northbrook Heights       10
40                                   Platt      198
41                             Poet's Hill        4
42                                Prospect       81
43                               Rail Side        3
44                          Regiment Creek        1
45                              Royal Road        7
46               Saint Mary's First Nation       25
47                 Saint Thomas University        1
48                              Sandyville        9
49                           Serenity Lane        2
50                       Shadowood Estates        5
51                              Silverwood       12
52                           Skyline Acrea       27
53                             South Devon       68
54                          Southwood Park       16
55                              Springhill        1
56                        Sunshine Gardens       10
57                                The Hill       44
58  The Hugh John Flemming Forestry Center        3
59             University Of New Brunswick       15
60                            Waterloo Row        9
61                          Wesbett / Case        1
62                              West Hills        5
63                 Williams / Hawkins Area       17
64                          Woodstock Road       41
65                         Youngs Crossing       16

In [55]:
address = 'Fredericton, Canada'

geolocator = Nominatim()
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Fredericton, New Brunswick is {}, {}.'.format(latitude, longitude))

/opt/conda/envs/Python36/lib/python3.6/site-packages/ipykernel/__main__.py:3: DeprecationWarning: Using Nominatim with the default "geopy/1.18.1" `user_agent` is strongly discouraged, as it violates Nominatim's ToS https://operations.osmfoundation.org/policies/nominatim/ and may possibly cause 403 and 429 HTTP errors. Please specify a custom `user_agent` with `Nominatim(user_agent="my-application")` or by overriding the default `user_agent`: `geopy.geocoders.options.default_user_agent = "my-application"`. In geopy 2.0 this will become an exception.
  app.launch_new_instance()


The geograpical coordinate of Fredericton, New Brunswick is 45.966425, -66.645813.


Geographical coordinates for Fredericton and New Brunswick are 45.966425, -66.645813 respectively

In [48]:
world_geo = r'world_countries.json' 

fredericton_1_map = folium.Map(location=[45.97, -66.65], width=700, height=500,zoom_start=10)

fredericton_1_map

In [84]:
fredericton_geo = r.json()

threshold_scale = np.linspace(CD['C_Count'].min(),CD['C_Count'].max(), 6,dtype=int)
threshold_scale = threshold_scale.tolist()
threshold_scale[-1] = threshold_scale[-1]+1

fredericton_1_map.choropleth(geo_data=fredericton_geo, data=CD,columns=['Neighb', 'C_Count'],
    key_on='feature.properties.Neighb', threshold_scale=threshold_scale,fill_color='YlOrRd', fill_opacity=0.7, 
    line_opacity=0.1, legend_name='Fredericton Neighbourhoods')

fredericton_1_map

NameError: name 'r' is not defined

<h3> Now unto Crime Types </h3>

In [85]:
CTD = df_data_1.groupby(['Crime_Type']).size().to_frame(name = 'Count').reset_index()
CTD

Crime_Type  Count
0                              4
1                   ARSON      5
2            ARSON BY NEG      1
3         ARSON-DAM.PROP.      4
4        B&E NON-RESIDNCE     51
5               B&E OTHER     58
6           B&E RESIDENCE    151
7        B&E STEAL FIREAR      3
8        MISCHIEF OBS USE      1
9        MISCHIEF TO PROP    246
10          MISCHIEF-DATA      2
11        MOTOR VEH THEFT     40
12       THEFT BIKE<$5000     63
13  THEFT FROM MV < $5000    356
14  THEFT FROM MV > $5000      5
15       THEFT OTH <$5000    458
16       THEFT OTH >$5000      9
17       THEFT OVER $5000      1
18       THEFT,BIKE>$5000      2

In [86]:
CTD.describe()

Count
count   19.000000
mean    76.842105
std    133.196706
min      1.000000
25%      2.500000
50%      5.000000
75%     60.500000
max    458.000000

In [87]:
CP = df_data_1.pivot_table(index='Neighbourhood', columns='Crime_Type', aggfunc=pd.Series.count, fill_value=0)
CP

City                     \
Crime_Type                                  ARSON ARSON BY NEG   
Neighbourhood                                                    
Barkers Point                             0     0            0   
Brookside                                 0     0            0   
Brookside Estates                         0     0            0   
Brookside Mini Home Park                  0     0            0   
College Hill                              0     2            0   
Colonial heights                          0     0            0   
Cotton Mill Creek                         0     0            0   
Diamond Street                            0     0            0   
Doak Road                                 0     0            0   
Douglas                                   0     0            0   
Downtown                                  0     1            0   
Dun's Crossing                            0     0            0   
Forest Hill                               0     0            0   
Fredericton South                         1     0            0   
Fulton Heights                            0     0            0   
Garden Creek                              0     0            0   
Garden Place                              0     0            0   
Gilridge Estates                          0     0            0   
Golf Club                                 0     0            0   
Grasse Circle                             1     0            0   
Greenwood Minihome Park                   0     0            0   
Hanwell North                             0     0            0   
Heron Springs                             0     0            0   
Highpoint Ridge                           0     0            0   
Kelly's Court Minihome Park               0     0            0   
Knob Hill                                 0     0            0   
Knowledge Park                            1     0            0   
Lian / Valcore                            0     0            0   
Lincoln                                   0     0            0   
Lincoln Heights                           0     0            0   
Main Street                               0     0            0   
Marysville                                0     1            0   
McKnight                                  0     0            0   
McLeod Hill                               0     0            0   
Monteith / Talisman                       0     0            0   
Montogomery / Prospect East               0     0            0   
Nashwaaksis                               0     0            0   
Nethervue Minihome Park                   0     0            0   
North Devon                               0     0            0   
Northbrook Heights                        0     0            0   
Platt                                     0     0            0   
Poet's Hill                               0     0            0   
Prospect                                  0     0            0   
Rail Side                                 0     0            0   
Regiment Creek                            0     0            0   
Royal Road                                0     0            0   
Saint Mary's First Nation                 0     0            0   
Saint Thomas University                   0     0            0   
Sandyville                                0     0            0   
Serenity Lane                             0     0            0   
Shadowood Estates                         0     0            0   
Silverwood                                0     0            0   
Skyline Acrea                             0     1            0   
South Devon                               0     0            1   
Southwood Park                            0     0            0   
Springhill                                0     0            0   
Sunshine Gardens                          0     0            0   
The Hill                                  0     0            0   
The Hugh John

<h3> Trying to Vizualize with a horizontal bar graph</h3>

In [96]:
import matplotlib
import matplotlib.pyplot as plt
CTD.plt(x='Crime_Type', y='Count', kind='barh')

AttributeError: 'property' object has no attribute '__name__'

<h3> Examining theft from vehicle less than $5k</h3>

In [97]:
VCD = df_data_1.loc[df_data_1['Crime_Type'] == 'THEFT FROM MV < $5000']

In [98]:
VCD

Neighbourhood  Crime_Code             Crime_Type  Ward  \
18              Fredericton South        2142  THEFT FROM MV < $5000     7   
19              Fredericton South        2142  THEFT FROM MV < $5000     7   
20              Fredericton South        2142  THEFT FROM MV < $5000     7   
21              Fredericton South        2142  THEFT FROM MV < $5000    12   
22              Fredericton South        2142  THEFT FROM MV < $5000    12   
23              Fredericton South        2142  THEFT FROM MV < $5000     7   
24              Fredericton South        2142  THEFT FROM MV < $5000     7   
25              Fredericton South        2142  THEFT FROM MV < $5000     7   
26              Fredericton South        2142  THEFT FROM MV < $5000    11   
27              Fredericton South        2142  THEFT FROM MV < $5000    11   
28              Fredericton South        2142  THEFT FROM MV < $5000    12   
29              Fredericton South        2142  THEFT FROM MV < $5000    12   
30              Fredericton South        2142  THEFT FROM MV < $5000     7   
51                  Barkers Point        2142  THEFT FROM MV < $5000     6   
52                  Barkers Point        2142  THEFT FROM MV < $5000     6   
53                  Barkers Point        2142  THEFT FROM MV < $5000     6   
54                  Barkers Point        2142  THEFT FROM MV < $5000     6   
55                  Barkers Point        2142  THEFT FROM MV < $5000     6   
56                  Barkers Point        2142  THEFT FROM MV < $5000     6   
57                  Barkers Point        2142  THEFT FROM MV < $5000     6   
58                  Barkers Point        2142  THEFT FROM MV < $5000     6   
100                    Sandyville        2142  THEFT FROM MV < $5000     5   
107                   South Devon        2142  THEFT FROM MV < $5000     4   
108                   South Devon        2142  THEFT FROM MV < $5000     4   
109                   South Devon        2142  THEFT FROM MV < $5000     4   
110                   South Devon        2142  THEFT FROM MV < $5000     4   
111                   South Devon        2142  THEFT FROM MV < $5000     4   
112                   South Devon        2142  THEFT FROM MV < $5000     4   
113                   South Devon        2142  THEFT FROM MV < $5000     4   
114                   South Devon        2142  THEFT FROM MV < $5000     4   
115                   South Devon        2142  THEFT FROM MV < $5000     4   
116                   South Devon        2142  THEFT FROM MV < $5000     4   
117                   South Devon        2142  THEFT FROM MV < $5000     4   
118                   South Devon        2142  THEFT FROM MV < $5000     4   
119                   South Devon        2142  THEFT FROM MV < $5000     4   
120                   South Devon        2142  THEFT FROM MV < $5000     4   
121                   South Devon        2142  THEFT FROM MV < $5000     4   
122                   South Devon        2142  THEFT FROM MV < $5000     4   
123                   South Devon        2142  THEFT FROM MV < $5000     4   
124                   South Devon        2142  THEFT FROM MV < $5000     4   
125                   South Devon        2142  THEFT FROM MV < $5000     4   
126                   South Devon        2142  THEFT FROM MV < $5000     4   
127                   South Devon        2142  THEFT FROM MV < $5000     4   
128                   South Devon        2142  THEFT FROM MV < $5000     4   
151                    Sandyville        2142  THEFT FROM MV < $5000     5   
156                     Knob Hill        2142  THEFT FROM MV < $5000     5   
165               Youngs Crossing        2142  THEFT FROM MV < $5000     4   
166               Youngs Crossing        2142  THEFT FROM MV < $5000     4   
167               Youngs Crossing        2142  THEFT FROM MV < $5000     4   
168               Youngs Crossing        2142  THEFT FROM MV < $5000     4   
169               Youngs Crossing        2

In [99]:
VCD1 = VCD.groupby(['Neighbourhood']).size().to_frame(name = 'Count').reset_index()

In [100]:
VCD1

Neighbourhood  Count
0                 Barkers Point      8
1             Brookside Estates      3
2                  College Hill     10
3              Colonial heights      6
4                Diamond Street      1
5                       Douglas      1
6                      Downtown     21
7                Dun's Crossing      9
8                   Forest Hill      8
9             Fredericton South     20
10               Fulton Heights     12
11                 Garden Creek      1
12                 Garden Place      3
13             Gilridge Estates      1
14                    Golf Club      5
15                Hanwell North      3
16                Heron Springs      2
17              Highpoint Ridge      4
18                    Knob Hill      1
19               Lian / Valcore      1
20                      Lincoln      1
21              Lincoln Heights     11
22                  Main Street     10
23                   Marysville     10
24                     McKnight      1
25                  McLeod Hill      2
26          Monteith / Talisman      3
27  Montogomery / Prospect East      3
28                  Nashwaaksis      9
29      Nethervue Minihome Park      1
30                  North Devon     17
31           Northbrook Heights      5
32                        Platt     40
33                  Poet's Hill      2
34                     Prospect     11
35                    Rail Side      2
36    Saint Mary's First Nation      1
37      Saint Thomas University      1
38                   Sandyville      3
39            Shadowood Estates      2
40                   Silverwood      2
41                Skyline Acrea     13
42                  South Devon     22
43               Southwood Park      7
44             Sunshine Gardens      7
45                     The Hill     11
46  University Of New Brunswick      4
47                 Waterloo Row      3
48      Williams / Hawkins Area      6
49               Woodstock Road     20
50              Youngs Crossing      6

In [101]:
VCD1.describe()

Count
count  51.000000
mean    6.980392
std     7.457855
min     1.000000
25%     2.000000
50%     4.000000
75%    10.000000
max    40.000000

In [102]:
VCD1.rename({'Platt': 'PLT'},inplace=True)
VCD1.rename(index=str, columns={'Neighbourhood':'Neighb','Count':'MC_Count'}, inplace=True)
VCD1

Neighb  MC_Count
0                 Barkers Point         8
1             Brookside Estates         3
2                  College Hill        10
3              Colonial heights         6
4                Diamond Street         1
5                       Douglas         1
6                      Downtown        21
7                Dun's Crossing         9
8                   Forest Hill         8
9             Fredericton South        20
10               Fulton Heights        12
11                 Garden Creek         1
12                 Garden Place         3
13             Gilridge Estates         1
14                    Golf Club         5
15                Hanwell North         3
16                Heron Springs         2
17              Highpoint Ridge         4
18                    Knob Hill         1
19               Lian / Valcore         1
20                      Lincoln         1
21              Lincoln Heights        11
22                  Main Street        10
23                   Marysville        10
24                     McKnight         1
25                  McLeod Hill         2
26          Monteith / Talisman         3
27  Montogomery / Prospect East         3
28                  Nashwaaksis         9
29      Nethervue Minihome Park         1
30                  North Devon        17
31           Northbrook Heights         5
32                        Platt        40
33                  Poet's Hill         2
34                     Prospect        11
35                    Rail Side         2
36    Saint Mary's First Nation         1
37      Saint Thomas University         1
38                   Sandyville         3
39            Shadowood Estates         2
40                   Silverwood         2
41                Skyline Acrea        13
42                  South Devon        22
43               Southwood Park         7
44             Sunshine Gardens         7
45                     The Hill        11
46  University Of New Brunswick         4
47                 Waterloo Row         3
48      Williams / Hawkins Area         6
49               Woodstock Road        20
50              Youngs Crossing         6

In [103]:
world_geo = r'world_countries.json' 

fredericton_c_map = folium.Map(location=[45.91, -66.65], width=850, height=650,zoom_start=10)

fredericton_c_map

 <h3>Motor Vehicle Crime less than $5k Count </h3>


In [104]:
fredericton_geo = r.json()
threshold_scale = np.linspace(VCD1['MC_Count'].min(), VCD1['MC_Count'].max(),6,dtype=int)
threshold_scale = threshold_scale.tolist()
threshold_scale[-1] = threshold_scale[-1]+1

fredericton_c_map.choropleth(geo_data=fredericton_geo,data=VCD1,columns=['Neighb', 'MC_Count'],key_on='feature.properties.Neighb',
    threshold_scale=threshold_scale, fill_color='YlOrRd',fill_opacity=0.7,line_opacity=0.1,legend_name='Fredericton Neighbourhoods')
fredericton_c_map

NameError: name 'r' is not defined

<h3> There is a high crime rate in Fredericton, Let's explore the cause of it</h3>

In [105]:

body = client_51343ec9129d4d20a45e6e5658b97f29.get_object(Bucket='ml1-donotdelete-pr-gbedlbdzfg4r4q',Key='_Donnes_dmographiques_du_secteur_de_recensement-6pge.csv')['Body']
# add missing __iter__ method, so pandas accepts body as file-like object
if not hasattr(body, "__iter__"): body.__iter__ = types.MethodType( __iter__, body )

df_data_2 = pd.read_csv(body)
df_data_2.head()


FID  OBJECTID       DBUID     DAUID  CDUID      CTUID  CTNAME     DBuid_1  \
0    1       501  1310024304  13100243   1310  3200002.0     2.0  1310024304   
1    2       502  1310032004  13100320   1310  3200010.0    10.0  1310032004   
2    3       503  1310017103  13100171   1310  3200014.0    14.0  1310017103   
3    4       504  1310018301  13100183   1310  3200012.0    12.0  1310018301   
4    5       505  1310022905  13100229   1310  3200007.0     7.0  1310022905   

   DBpop2011  DBtdwell20  DBurdwell2  Shape_Leng  Shape_Area  CTIDLINK  \
0         60          25          22    0.007462    0.000003   3200002   
1         15           3           3    0.009008    0.000003   3200010   
2          0           0           0    0.010602    0.000007   3200014   
3        108          60          50    0.039599    0.000068   3200012   
4        129          47          44    0.011833    0.000005   3200007   

    Shape__Area  Shape__Length  
0  5.013915e+04     929.421025  
1  5.914999e+04    1211.100736  
2  1.206995e+05    1451.322317  
3  1.204640e+06    5615.379455  
4  9.439736e+04    1507.917222

<h3> Let's assume one of the major causes of crime in the city is due to population density </h3>

In [77]:

world_geo = r'world_countries.json' 
fredericton_d_map = folium.Map(location=[45.94, -66.63], width=1000, height=700,zoom_start=10)
fredericton_d_map

threshold_scale = np.linspace(df_data_2['DBpop2011'].min(),df_data_2['DBpop2011'].max(),6,dtype=int)
threshold_scale = threshold_scale.tolist()
threshold_scale[-1] = threshold_scale[-1]+1

fredericton_d_map.choropleth(geo_data=world_geo,data=df_data_2,columns=['OBJECTID','DBpop2011'],key_on='feature.properties.OBJECTID',
    threshold_scale=threshold_scale,fill_color='PuBuGn',fill_opacity=0.7, line_opacity=0.1,legend_name='Fredericton Population Density')
fredericton_d_map

FileNotFoundError: [Errno 2] No such file or directory: 'world_countries.json'

<h3> Now we search specific locations in the Fredericton city</h3>

In [106]:

body = client_51343ec9129d4d20a45e6e5658b97f29.get_object(Bucket='ml1-donotdelete-pr-gbedlbdzfg4r4q',Key='Fredericton Locations.xlsx')['Body']
# add missing __iter__ method, so pandas accepts body as file-like object
if not hasattr(body, "__iter__"): body.__iter__ = types.MethodType( __iter__, body )

df_data_0 = pd.read_excel(body)
df_data_0.head()


Unnamed: 0                     Location   Latitude  Longitude
0           0               Knowledge Park  45.931143 -66.652700
1           1         Fredericton Downtown  45.963026 -66.383550
2           2             Fredericton Hill  45.948512 -66.656045
3           3                  Nashwaaksis  45.983382 -66.644856
4           4  University of New Brunswick  45.948121 -66.641406

In [109]:

body = client_51343ec9129d4d20a45e6e5658b97f29.get_object(Bucket='ml1-donotdelete-pr-gbedlbdzfg4r4q',Key='Fredericton Locations.xlsx')['Body']
# add missing __iter__ method, so pandas accepts body as file-like object
if not hasattr(body, "__iter__"): body.__iter__ = types.MethodType( __iter__, body )

df_data_0 = pd.read_excel(body)
df_data_0.head()

Unnamed: 0                     Location   Latitude  Longitude
0           0               Knowledge Park  45.931143 -66.652700
1           1         Fredericton Downtown  45.963026 -66.383550
2           2             Fredericton Hill  45.948512 -66.656045
3           3                  Nashwaaksis  45.983382 -66.644856
4           4  University of New Brunswick  45.948121 -66.641406

In [110]:
for lat, lng, point in zip(df_data_0['Latitude'], df_data_0['Longitude'], df_data_0['Location']):
    label = '{}'.format(point)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker([lat, lng],radium=1,popup=label,color='blue',fill=True,fill_color='#3186cc',fill_opacity=0.7,
        parse_html=False).add_to(fredericton_c_map)
fredericton_c_map

<h3>Now we are going to explore Fredericton Neighbourhood</h3>

In [114]:
CLIENT_ID = 'SRRJDG5S5OZZZNBA4A4I4CMIEEIMQRDBUTG1Y2EI23JBE05O' 
CLIENT_SECRET = 'L3VNPXUVP3Q0Y4TLDQKXI3ASX431FNKEBDDVSR35LNJKLYDC' 
VERSION = '20200629' 

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: SRRJDG5S5OZZZNBA4A4I4CMIEEIMQRDBUTG1Y2EI23JBE05O
CLIENT_SECRET:L3VNPXUVP3Q0Y4TLDQKXI3ASX431FNKEBDDVSR35LNJKLYDC


In [115]:
def getNearbyVenues(names, latitudes, longitudes, radius=800, LIMIT=100):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng,            
            v['venue']['name'], 
            v['venue']['id'],
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Location', 
                  'Location Latitude', 
                  'Location Longitude', 
                  'Venue',
                  'Venue id',                
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category'        
                   ]
    return(nearby_venues)

In [117]:
FDV = getNearbyVenues(names=df_data_0['Location'],
                                   latitudes=df_data_0['Latitude'],
                                   longitudes=df_data_0['Longitude']
                                  )

Knowledge Park
Fredericton Downtown
Fredericton Hill
Nashwaaksis
University of New Brunswick
Devon
New Maryland
Marysville
Skyline Acres
Hanwell


In [118]:
FDV.shape

(62, 8)

In [119]:
FDV

Location  Location Latitude  Location Longitude  \
0                Knowledge Park          45.931143          -66.652700   
1                Knowledge Park          45.931143          -66.652700   
2                Knowledge Park          45.931143          -66.652700   
3                Knowledge Park          45.931143          -66.652700   
4                Knowledge Park          45.931143          -66.652700   
5                Knowledge Park          45.931143          -66.652700   
6                Knowledge Park          45.931143          -66.652700   
7                Knowledge Park          45.931143          -66.652700   
8                Knowledge Park          45.931143          -66.652700   
9                Knowledge Park          45.931143          -66.652700   
10               Knowledge Park          45.931143          -66.652700   
11               Knowledge Park          45.931143          -66.652700   
12               Knowledge Park          45.931143          -66.652700   
13               Knowledge Park          45.931143          -66.652700   
14             Fredericton Hill          45.948512          -66.656045   
15             Fredericton Hill          45.948512          -66.656045   
16             Fredericton Hill          45.948512          -66.656045   
17             Fredericton Hill          45.948512          -66.656045   
18                  Nashwaaksis          45.983382          -66.644856   
19                  Nashwaaksis          45.983382          -66.644856   
20                  Nashwaaksis          45.983382          -66.644856   
21                  Nashwaaksis          45.983382          -66.644856   
22                  Nashwaaksis          45.983382          -66.644856   
23                  Nashwaaksis          45.983382          -66.644856   
24  University of New Brunswick          45.948121          -66.641406   
25  University of New Brunswick          45.948121          -66.641406   
26  University of New Brunswick          45.948121          -66.641406   
27  University of New Brunswick          45.948121          -66.641406   
28  University of New Brunswick          45.948121          -66.641406   
29  University of New Brunswick          45.948121          -66.641406   
30  University of New Brunswick          45.948121          -66.641406   
31  University of New Brunswick          45.948121          -66.641406   
32  University of New Brunswick          45.948121          -66.641406   
33  University of New Brunswick          45.948121          -66.641406   
34                        Devon          45.968802          -66.622738   
35                        Devon          45.968802          -66.622738   
36                        Devon          45.968802          -66.622738   
37                        Devon          45.968802          -66.622738   
38                        Devon          45.968802          -66.622738   
39                        Devon          45.968802          -66.622738   
40                        Devon          45.968802          -66.622738   
41                        Devon          45.968802          -66.622738   
42                        Devon          45.968802          -66.622738   
43                        Devon          45.968802          -66.622738   
44                 New Maryland          45.892795          -66.683673   
45                 New Maryland          45.892795          -66.683673   
46                 New Maryland          45.892795          -66.683673   
47                 New Maryland          45.892795          -66.683673   
48                   Marysville          45.978913          -66.589491   
49                   Marysville          45.978913          -66.589491   
50                   Marysville          45.978913          -66.589491   
51                   Marysville          45.978913          -66.589491   
52                   Marysville          45.978913          -66.589491   
53                Skyline A

In [120]:
print('The town has {} unique venue categories.'.format(len(FDV['Venue Category'].unique())))

The town has 39 unique venue categories.


In [121]:
print('The town has {} unique venues.'.format(len(FDV['Venue id'].unique())))

The town has 61 unique venues.


In [122]:
UN = FDV.groupby('Location').nunique('Venue Category')
UN

Location  Location Latitude  Location Longitude  \
Location                                                                       
Devon                               1                  1                   1   
Fredericton Hill                    1                  1                   1   
Hanwell                             1                  1                   1   
Knowledge Park                      1                  1                   1   
Marysville                          1                  1                   1   
Nashwaaksis                         1                  1                   1   
New Maryland                        1                  1                   1   
Skyline Acres                       1                  1                   1   
University of New Brunswick         1                  1                   1   

                             Venue  Venue id  Venue Latitude  Venue Longitude  \
Location                                                                        
Devon                           10        10              10               10   
Fredericton Hill                 4         4               4                4   
Hanwell                          4         4               4                4   
Knowledge Park                  14        14              14               14   
Marysville                       5         5               5                5   
Nashwaaksis                      6         6               6                6   
New Maryland                     4         4               4                4   
Skyline Acres                    5         5               5                5   
University of New Brunswick      9        10              10               10   

                             Venue Category  
Location                                     
Devon                                    10  
Fredericton Hill                          4  
Hanwell                                   4  
Knowledge Park                           11  
Marysville                                5  
Nashwaaksis                               6  
New Maryland                              4  
Skyline Acres                             5  
University of New Brunswick               9

<h3> We want to analyze the respective locations</h3>

In [124]:

FT = pd.get_dummies(FDV[['Venue Category']], prefix="", prefix_sep="")

# adding neighbourhood column back to dataframe
FT['Location'] = FDV['Location'] 

# moving neighbourhood column to the first column
FC = [FT.columns[-1]] + list(FT.columns[:-1])
FT = FT[FC]

FT.head()

Location  Arts & Crafts Store  Bakery  Bar  Baseball Field  \
0  Knowledge Park                    0       0    0               0   
1  Knowledge Park                    0       0    0               0   
2  Knowledge Park                    0       0    0               0   
3  Knowledge Park                    0       0    0               0   
4  Knowledge Park                    1       0    0               0   

   Baseball Stadium  Basketball Court  Clothing Store  Coffee Shop  \
0                 0                 0               0            0   
1                 0                 0               0            0   
2                 0                 0               0            0   
3                 0                 0               0            0   
4                 0                 0               0            0   

   Convenience Store  Department Store  Discount Store  Fast Food Restaurant  \
0                  0                 0               1                     0   
1                  0                 0               0                     0   
2                  0                 0               0                     0   
3                  0                 0               0                     0   
4                  0                 0               0                     0   

   Furniture / Home Store  Grocery Store  Gym  Gym / Fitness Center  \
0                       0              0    0                     0   
1                       1              0    0                     0   
2                       0              0    0                     0   
3                       0              0    0                     0   
4                       0              0    0                     0   

   Hockey Arena  Hotel  Ice Cream Shop  Insurance Office  Laundry Service  \
0             0      0               0                 0                0   
1             0      0               0                 0                0   
2             0      0               0                 0                0   
3             0      0               0                 0                0   
4             0      0               0                 0                0   

   Liquor Store  Martial Arts Dojo  Park  Pet Store  Pharmacy  Pizza Place  \
0             0                  0     0          0         0            0   
1             0                  0     0          0         0            0   
2             0                  0     0          1         0            0   
3             0                  0     0          0         0            0   
4             0                  0     0          0         0            0   

   Pool  Pub  Rental Service  Restaurant  Sandwich Place  Seafood Restaurant  \
0     0    0               0           0               0                   0   
1     0    0               0           0               0                   0   
2     0    0               0           0               0                   0   
3     0    0               0           1               0                   0   
4     0    0               0           0               0                   0   

   Shoe Store  Smoke Shop  Spa  Sports Bar  Tennis Court  Trail  
0           0           0    0           0             0      0  
1           0           0    0           0             0      0  
2           0           0    0           0             0      0  
3           0           0    0           0             0      0  
4           0           0    0           0             0      0

In [125]:
FT.shape

(62, 40)

<h3> Now, wea re grouping each row by mean and location </h3>

In [126]:
FG = FT.groupby('Location').mean().reset_index()
FG

Location  Arts & Crafts Store  Bakery   Bar  \
0                        Devon             0.000000     0.0  0.00   
1             Fredericton Hill             0.000000     0.0  0.25   
2                      Hanwell             0.000000     0.0  0.00   
3               Knowledge Park             0.071429     0.0  0.00   
4                   Marysville             0.000000     0.0  0.00   
5                  Nashwaaksis             0.000000     0.0  0.00   
6                 New Maryland             0.000000     0.0  0.00   
7                Skyline Acres             0.000000     0.0  0.00   
8  University of New Brunswick             0.000000     0.1  0.10   

   Baseball Field  Baseball Stadium  Basketball Court  Clothing Store  \
0            0.10               0.0               0.0        0.000000   
1            0.00               0.0               0.0        0.000000   
2            0.00               0.0               0.0        0.000000   
3            0.00               0.0               0.0        0.071429   
4            0.00               0.2               0.0        0.000000   
5            0.00               0.0               0.0        0.000000   
6            0.25               0.0               0.0        0.000000   
7            0.20               0.0               0.0        0.200000   
8            0.00               0.0               0.1        0.000000   

   Coffee Shop  Convenience Store  Department Store  Discount Store  \
0         0.10                0.0          0.000000        0.000000   
1         0.00                0.0          0.000000        0.000000   
2         0.25                0.0          0.000000        0.000000   
3         0.00                0.0          0.142857        0.071429   
4         0.20                0.2          0.000000        0.000000   
5         0.00                0.0          0.000000        0.000000   
6         0.00                0.0          0.000000        0.000000   
7         0.00                0.0          0.000000        0.000000   
8         0.20                0.0          0.000000        0.000000   

   Fast Food Restaurant  Furniture / Home Store  Grocery Store       Gym  \
0              0.100000                0.000000            0.1  0.000000   
1              0.000000                0.000000            0.0  0.250000   
2              0.000000                0.000000            0.0  0.000000   
3              0.071429                0.214286            0.0  0.000000   
4              0.000000                0.000000            0.0  0.000000   
5              0.000000                0.000000            0.0  0.166667   
6              0.250000                0.000000            0.0  0.000000   
7              0.000000                0.000000            0.0  0.000000   
8              0.000000                0.000000            0.1  0.000000   

   Gym / Fitness Center  Hockey Arena  Hotel  Ice Cream Shop  \
0                   0.0           0.0    0.0             0.1   
1                   0.0           0.0    0.0             0.0   
2                   0.0           0.0    0.0             0.0   
3                   0.0           0.0    0.0             0.0   
4                   0.0           0.0    0.0             0.0   
5                   0.0           0.0    0.0             0.0   
6                   0.0           0.0    0.0             0.0   
7                   0.0           0.2    0.2             0.0   
8                   0.1           0.0    0.0             0.0   

   Insurance Office  Laundry Service  Liquor Store  Martial Arts Dojo  Park  \
0              0.00             0.00           0.1               0.00  0.00   
1              0.00             0.00           0.0               0.00  0.25   
2              0.00             0.25           0.0               0.00  0.00   
3              0.00             0.00           0.0               0.00  0.00   
4              0.00             0.00           0.0               0.00  0.20   
5            

In [127]:
FG.shape

(9, 40)

<h3> Printing each location with top 5 most common venues </h3>

In [141]:
TV = 5

for hood in FG['Location']:
    print("----"+hood+"----")
    temp = FG[FG['Location'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(TV))
    print('\n')

----Devon----
                  venue  freq
0  Fast Food Restaurant   0.1
1           Coffee Shop   0.1
2              Pharmacy   0.1
3        Ice Cream Shop   0.1
4           Pizza Place   0.1


----Fredericton Hill----
                 venue  freq
0                  Bar  0.25
1                 Park  0.25
2             Pharmacy  0.25
3                  Gym  0.25
4  Arts & Crafts Store  0.00


----Hanwell----
                 venue  freq
0       Rental Service  0.25
1             Pharmacy  0.25
2          Coffee Shop  0.25
3      Laundry Service  0.25
4  Arts & Crafts Store  0.00


----Knowledge Park----
                    venue  freq
0  Furniture / Home Store  0.21
1        Department Store  0.14
2     Arts & Crafts Store  0.07
3                     Spa  0.07
4              Shoe Store  0.07


----Marysville----
               venue  freq
0               Park   0.2
1   Baseball Stadium   0.2
2           Pharmacy   0.2
3        Coffee Shop   0.2
4  Convenience Store   0.2


----Nashwaa

<h3> Getting into Pandas Data frame from here </h3>

In [142]:
def return_most_common_venues(row, TV):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:TV]

In [143]:
TV = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Location']
for ind in np.arange(TV):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
location_venues_sorted = pd.DataFrame(columns=columns)
location_venues_sorted['Location'] = FG['Location']

for ind in np.arange(FG.shape[0]):
    location_venues_sorted.iloc[ind, 1:] = return_most_common_venues(FG.iloc[ind, :], TV)

location_venues_sorted

Location   1st Most Common Venue 2nd Most Common Venue  \
0                        Devon          Ice Cream Shop         Grocery Store   
1             Fredericton Hill                     Bar                   Gym   
2                      Hanwell         Laundry Service           Coffee Shop   
3               Knowledge Park  Furniture / Home Store      Department Store   
4                   Marysville        Baseball Stadium                  Park   
5                  Nashwaaksis                   Trail            Sports Bar   
6                 New Maryland        Insurance Office        Baseball Field   
7                Skyline Acres            Tennis Court                 Hotel   
8  University of New Brunswick             Coffee Shop                   Pub   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0            Smoke Shop        Baseball Field    Seafood Restaurant   
1                  Park              Pharmacy                 Trail   
2        Rental Service              Pharmacy                 Trail   
3                 Trail            Restaurant        Clothing Store   
4           Coffee Shop     Convenience Store              Pharmacy   
5                   Gym        Sandwich Place              Pharmacy   
6     Martial Arts Dojo  Fast Food Restaurant        Discount Store   
7          Hockey Arena        Baseball Field        Clothing Store   
8                Bakery                   Bar  Gym / Fitness Center   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0          Liquor Store  Fast Food Restaurant           Coffee Shop   
1        Discount Store  Gym / Fitness Center         Grocery Store   
2        Discount Store  Gym / Fitness Center                   Gym   
3        Discount Store  Fast Food Restaurant             Pet Store   
4                 Trail  Fast Food Restaurant  Gym / Fitness Center   
5                  Pool     Convenience Store         Grocery Store   
6          Hockey Arena  Gym / Fitness Center                   Gym   
7        Discount Store  Gym / Fitness Center                   Gym   
8      Basketball Court         Grocery Store            Restaurant   

    9th Most Common Venue  10th Most Common Venue  
0             Pizza Place                Pharmacy  
1  Furniture / Home Store    Fast Food Restaurant  
2           Grocery Store  Furniture / Home Store  
3     Arts & Crafts Store                     Spa  
4                     Gym           Grocery Store  
5  Furniture / Home Store    Fast Food Restaurant  
6           Grocery Store  Furniture / Home Store  
7           Grocery Store  Furniture / Home Store  
8                    Park          Discount Store

<h2> Clustering Fredericton Location into 5</h2>

In [144]:
k = 5

FGC = FG.drop('Location', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=k, random_state=0).fit(FGC)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([1, 4, 0, 1, 1, 4, 3, 2, 1], dtype=int32)

<h3>I am creating a new dataframe together with cluster and the top 10 venues for all Locations</h3>

In [145]:
FM = df_data_0

# add clustering labels
FM['Cluster Labels'] = kmeans.labels_

# merge fredericton_grouped with location df to add latitude/longitude for each location
FM = FM.join(location_venues_sorted.set_index('Location'), on='Location')

FM

ValueError: Length of values does not match length of index